_source/_posts/2022-09-26-micro-frontends-java-microservices.adoc

```diff
@@ -0,0 +1,625 @@
+
+image::{% asset_path 'blog/micro-frontends-java/jhipster-microservices.png' %}[alt=JHipster microservices architecture,width=800,align=center]
+
+This microservices architecture had one major flaw: the UI for it is a monolith, with all its files on the gateway. This isn't good from a loose-coupling point of view because changes in a microservice might require changes in the UI. Instead of being able to independently deploy the microservice, you have to deploy the gateway too.
+
+Today, I'm proud to show you how to you can solve this problem with micro frontends. JHipster recently added support for micro-frontends. Microfrontends provide a way for you to remotely load and execute code at runtime so your microservice's UI can live in the same artifact, without being coupled to the gateway!
+
+image::{% asset_path 'blog/micro-frontends-java/jhipster-microfrontends.png' %}[alt=JHipster microfrontends architecture,width=800,align=center]
+
+You might notice I spelled micro frontends three different ways in the previous paragraph. The current literature is all over the place on this one!
+
+// https://twitter.com/mraible/status/1499098253920460802
+++++
+{% twitter 1499098253920460802 %}
+++++
+
+I'm going to use "micro frontends" for the remainder of this post since that's what https://twitter.com/thecamjackson[Cam Jackson] used in his https://martinfowler.com/articles/micro-frontends.html[Micro Frontends article] on Martin Fowler's blog.
+
+== A quick introduction to Module Federation
+
+Webpack's https://webpack.js.org/concepts/module-federation/[Module Federation] is one of the best known implementations for micro frontends. Its https://webpack.js.org/plugins/module-federation-plugin[`ModuleFederationPlugin`] allows a build to provide or consume modules with other independent builds at runtime. It even allows you to share libraries between frontends to reduce the size of remote bundles.
+
+https://twitter.com/ScriptedAlchemy[Zack Jackson] is the creator of Module Federation and recently https://twitter.com/ScriptedAlchemy/status/1564411584851505153[collaborated] with https://twitter.com/ManfredSteyer[Manfred Steyer] to create Native Federation. This means you can use micro frontend concepts with any build tool, not just webpack.
+
+// https://twitter.com/ManfredSteyer/status/1564312149580582912
+++++
+{% twitter 1564312149580582912 %}
+++++
+
+== Why should Java developers care?
+
+I think micro frontends an excellent architecture concept. With microservices, you weave everything together on the backend with protocols like HTTP, gRPC, and GraphQL. With micro frontends, it's all HTTP. You can literally see your app get stitched together by watching your browser's network console and seeing remote modules load.
```

Comment:
I wouldn't call GraphQL a protocol. Isn't it a query language and more like REST rather than HTTP or gRPC?

Reply:
Fixed.